src/java.base/share/classes/jdk/internal/misc/CDS.java

```diff
@@ -109,11 +110,21 @@ public static boolean isDumpingStaticArchive() {
      */
     public static native long getRandomSeedForDumping();
 
+    /**
+     * Checks if the lambda form invoker is archivable. This is to avoid RuntimeException
+     * during CDS archive dumping.
+     */
+    private static boolean isInvokerArchivable(MethodType mt, String holder) {
+        final int lastParam = mt.parameterCount() - 1;
+        return ((mt.parameterCount() >= 2 && mt.parameterType(0) == Object.class
+                && mt.parameterType(lastParam) == Object.class) || isValidHolderName(holder));
+    }
+
     /**
      * log lambda form invoker holder, name and method type
      */
-    public static void traceLambdaFormInvoker(String prefix, String holder, String name, String type) {
-        if (isDumpingClassList) {
+    public static void traceLambdaFormInvoker(String prefix, String holder, String name, String type, MethodType mt) {
+        if (isDumpingClassList && isInvokerArchivable(mt, holder)) {
```

Comment:
Is this check still needed after you added the filtering in GenerateJLIClassesHelper.java?

Since the logic here is not 100% the same as the logic in GenerateJLIClassesHelper.java, we might be filtering more things than necessary.

Reply:
I did some sanity tests without the changes in CDS.java and it seems to work.
I'll do more testing before pushing another commit.